### ISSUE TITLE
S2583 FP: Variable change not detected inside a loop

### ISSUE THREAD

**Issue by paulhickman-a365** [SonarSource/sonar-dotnet#5601]
### Description

The following method when executed returns true, but SonarAnalyser gives a warning on the last "if" statement that the expression is always false.

If you delete the 2nd last "if" the warning goes away

### Repro steps

````c#
namespace ConsoleApp1
{
    public static class Test
    {
        public static bool Run()
        {
            int pos = 0;
            bool foundA = false;
            bool readA = false;
            string test = "ab";

            while (pos < test.Length)
            {
                if (test[pos] == 'a')
                {
                    foundA = true;
                    readA = true;
                }
                else if (readA)
                {
                    readA = false;
                }
                pos++;
            }

            if (readA)
            {
                return false;
            }

            if (foundA)
            {
                return true;
            }
            return false;
        }
    }
}

````
### Expected behavior

No warnings

### Actual behavior

On the line ````if (foundA)````
Warning	S2583	Change this condition so that it does not always evaluate to 'false'; some subsequent code is never executed.	

### Known workarounds

Supress the warning

### Related information

.NET 6, Sonar Analyser.CSharp 8.38.0.46746

**Comment by pavel-mikula-sonarsource** [comment]
Hi @paulhickman-a365,

Thank you for reporting this case. I can confirm it as False Positive.

As a workaround, you can also refactor your code like this to avoid the issue:
```
return foundA;
```
Further refactoring will raise the issue again:
```
return foundA && !readA;
```

**Comment by markusschaber** [comment]
We also have several occurences of this.
